bzzeth/bzzeth.go

```diff
@@ -79,6 +79,7 @@ func (b *BzzEth) Run(p *p2p.Peer, rw p2p.MsgReadWriter) error {
 	// perform handshake and register if peer serves headers
 	handshake, err := bp.Handshake(context.TODO(), Handshake{ServeHeaders: true}, nil)
 	if err != nil {
+		log.Error("Shutting down bzzeth protocol.", "peer", bp, "reason", err)
```

Comment:
Error should not be logged and passed, just one of those.

Reply:
This (and the next comment) were the things I wanted to discuss.  I see 3 scenarios:
1. Log some context and message here and return nil
2. Log some context here and return a new specific error that will be seen in events in tests
3. Add more context too the error here and return an error. I am not sure if this error will be logged anywhere.

Reply:
Scenario 3. is the one that I think is the correct one. We now have BreakError which is to make the peer drop, and any other error will be at least logged. If the error is logged and returned, it is easy to make the mistake to log the error twice. If the error is just passed, it is up to the caller to decide what to do, and the function that returns the error does not have to worry about it.

Reply:
I agree, from the architecture point of view. I am just worried about the `BreakError`s, which won't be logged in this case, right? This is probably the place that has enough information about what will happen to the specific protocol (ex. bzzeth, stream, etc...).

Reply:
If there is a BreakError, then it should be logged as error before the peer is dropped.

Reply:
The peer is actually dropped when the run function returns. This is why I was logging it here, as the drop is actually happening when it's propagated with end of the run function, in this  case  when `BzzEth.Run()` returns. This is why I was logging it here. The error returned by `BzzEth.Run()` is only used for events and testDisconnect. So, I see 3 options if we don't won't to log and return err message here: 
1. Use `peer.Drop()` on error in `p2p.Protocol` and log BreakErrors there before drop. No need to return specific errors in `p2p.Protocol.Run()` then at all maybe?
2. Log the specific info here, and return different error to be shown in events and drop

I think the main question is, weather we need to return specific errors with `p2p.Protocol.Run()` and weather we want to rely on finishing `Run()` functions to drop a peer or do we want to drop a peer manually.

Reply:
Cool, as we discussed, option 1. would be a better compromise.

Reply:
Ok, I have switched back to option 1. I don't like the fact that we lose a bit of control over the async errors. It's hard to check what caused protocol to stop, except that there is a log message. Of course, we could add the functionality in p2p.protocol to remember the drop error, and expose the API to check it. I could implement it, and use it in some tests. But maybe we can think about it in the future, and leave this as is for now, a we tried both approaches and it should not be to hard to switch back and forth between them.